File: GherkinEditor/GherkinEditor/Model/BDD/BDDStepArg.cs
using System.Text.RegularExpressions;

namespace CucumberCpp
{
    enum BDDStepArgType { TableArg, DocStringArg, IntArg, FloatArg, StringArg, TableColumnArg }

    class BDDStepArg
    {
        public const string TableArg = "$table$";
        public const string DocStringArg = "$doc$";
        public BDDStepArg(string arg)
        {
            RegexPattern = BDDUtil.StringRegex; // default is a string argument
            ArgText = arg;
            DecideArgType(arg);
        }

        public int ArgIndex { get; set; }
        public string RegexPattern { get; set; }
        public BDDStepArgType ArgType { get; set; }
        string ArgText { get; set; }

        public string StepParam
        {
            get
            {
                switch (ArgType)
                {
                    case BDDStepArgType.TableArg:
                        return "GherkinTable&, table";
                    case BDDStepArgType.TableColumnArg:
                        return "GherkinRow&, row";
                    case BDDStepArgType.DocStringArg:
                        return "std::wstring, docStr";
                    case BDDStepArgType.IntArg:
                        return "int, num" + ArgIndex.ToString();
                    case BDDStepArgType.FloatArg:
                        return "double, num" + ArgIndex.ToString();
                    case BDDStepArgType.StringArg:
                        return "std::wstring, str" + ArgIndex.ToString();
                    default:
                        return "std::wstring, str" + ArgIndex.ToString();
                }
            }
        }

        public string StepParamTypeName
        {
            get
            {
                switch (ArgType)
                {
                    case BDDStepArgType.TableArg:
                        return "GherkinTable&";
                    case BDDStepArgType.TableColumnArg:
                        return "GherkinRow&";
                    case BDDStepArgType.DocStringArg:
                        return "wstring";
                    case BDDStepArgType.FloatArg:
                        return "double";
                    case BDDStepArgType.StringArg:
                        return "wstring";
                    default:
                        return "wstring";
                }
            }
        }

        public string StepFunctionPlaceHolder
        {
            get
            {
                switch (ArgType)
                {
                    case BDDStepArgType.TableArg:
                        return "_T_";
                    case BDDStepArgType.TableColumnArg:
                        string tableColumnName = ArgText.Substring(1, ArgText.Length - 2);
                        return BDDUtil.MakeIdentifier(tableColumnName);
                    case BDDStepArgType.DocStringArg:
                        return "_S_";
                    case BDDStepArgType.IntArg:
                        return "_I_";
                    case BDDStepArgType.FloatArg:
                        return "_D_";
                    case BDDStepArgType.StringArg:
                        return "_S_";
                    default:
                        return "";
                }
            }
        }

        void DecideArgType(string arg)
        {
            if (arg == TableArg)
            {
                ArgType = BDDStepArgType.TableArg;
            }
            else if (arg == DocStringArg)
            {
                ArgType = BDDStepArgType.DocStringArg;
            }
            else if ((arg.Length == 0) || (arg[0] == '\"'))
            {
                ArgType = BDDStepArgType.StringArg;
                RegexPattern = BDDUtil.StringRegex;
            }
            else if ((arg[0] == '<') && (arg[arg.Length - 1] == '>'))
            {
                ArgType = BDDStepArgType.TableColumnArg;
                RegexPattern = arg;
            }
            else if (!arg.Contains("."))
            {
                ArgType = BDDStepArgType.IntArg;
                RegexPattern = BDDUtil.IntRegex;
            }
            else
            {
                ArgType = BDDStepArgType.FloatArg;
                RegexPattern = BDDUtil.FloatRegex;
            }
        }
    }
}

File: TestModelForCPP/BDDStepArg.cpp
#include "RegexSubstituter.h"
#include "StringUtility.h"
#include "BDDStepArg.h"

using namespace std;
using namespace bdd;

const std::wstring BDDStepArg::TableArg(L"$table$");
const std::wstring BDDStepArg::DocStringArg(L"$doc$");

BDDStepArg::BDDStepArg(wstring arg) :
	ArgIndex(0),
	m_RegexPattern(RegexSubstituter::StringRegex()), // default is a string argument
	m_ArgText(arg)
{
    DeduceArgType(arg);
}

wstring BDDStepArg::GetStepParam()
{
    switch (m_ArgType)
    {
    case BDDStepArgType::TableArg:
        return L"GherkinTable&, table";
    case BDDStepArgType::TableColumnArg:
        return L"GherkinRow&, row";
    case BDDStepArgType::DocStringArg:
        return L"std::wstring, docStr";
    case BDDStepArgType::IntArg:
        return wstring(L"int, num") + StringUtility::itows(ArgIndex);
    case BDDStepArgType::FloatArg:
        return wstring(L"double, num") + StringUtility::itows(ArgIndex);
    case BDDStepArgType::StringArg:
        return wstring(L"std::wstring, str") + StringUtility::itows(ArgIndex);
    default:
        return wstring(L"std::wstring, str") + StringUtility::itows(ArgIndex);
    }
}

void BDDStepArg::DeduceArgType(wstring arg)
{
    if (arg == TableArg)
    {
        m_ArgType = BDDStepArgType::TableArg;
    }
    else if (arg == DocStringArg)
    {
        m_ArgType = BDDStepArgType::DocStringArg;
    }
    else if ((arg.length() == 0) || (arg[0] == L'\"'))
    {
        m_ArgType = BDDStepArgType::StringArg;
        m_RegexPattern = RegexSubstituter::StringRegex();
    }
    else if ((arg[0] == L'<') && (arg[arg.length() - 1] == L'>'))
    {
        m_ArgType = BDDStepArgType::TableColumnArg;
        m_RegexPattern = arg;
    }
    else if (arg.find(L'.') == std::wstring::npos)
    {
        m_ArgType = BDDStepArgType::IntArg;
        m_RegexPattern = RegexSubstituter::IntRegex();
    }
    else
    {
        m_ArgType = BDDStepArgType::FloatArg;
        m_RegexPattern = RegexSubstituter::FloatRegex();
    }
}
